pythonPackages.slither-analyzer: fixed dependencies (+1 package)

jerengie opened the pull request:
###### Motivation for this change
`pythonPackage.slither-analyzer` can not be built at this moment.
This is due to missing / unlisted dependencies in the package formula.
This PR provides a fix for this and updates the formula to include the existing `pythonPackage.pysha3` package.
Additionally, another dependency, `crytic-compile`, is packaged as `pythonPackage.crytic-compile` in this PR.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

jonringer commented on pkgs/development/python-modules/crytic-compile/default.nix:
```diff
@@ -0,0 +1,39 @@
+  postFixup = ''
+    wrapProgram $out/bin/crytic-compile \
+      --prefix PATH : "${lib.makeBinPath [ solc ]}"
+  '';
```
does the package also shell out to solc? if it does, then this wont work if someone imports this as a package.

jerengie replied:
If used as a standalone tool, yes. 
I am not aware of any nix package that depends on this besides `slither-analyzer`, which also includes this `postFixup`.

asymmetric replied:
> does the package also shell out to solc? if it does, then this wont work if someone imports this as a package.

@jonringer why is that?

jonringer replied:
because you only wrap the executable, and the python package package has no knowledge of it.

If i had another package which has `propagatedBuildInputs = [ cryptic-compile ];` as a dependenccy, then when I call `$ my_package_exe` and it eventually calls cryptic-compile, which will do something to the effect of `subprocess.check_call("solc" ...)`, then solc won't be present.

If you want the package to do the right thing, you need to patch the call sights inside of only wrapping the entry_points.

jonringer commented on maintainers/maintainer-list.nix:
```diff
@@ -3349,6 +3349,12 @@
+    github = "jerengie";
+    githubID = 1221065;
+    name = "Jens-Rene Giesen";
+  };
```
this should be a separate commit as well:
```
maintainers: add jerengie
```

jerengie replied:
Thank you, I corrected this.

jonringer reviewed (commented):
there should be 3 commits:
```
maintainers: add jerengie
pythonPackages.slither-analyzer: fix dependencies
pythonPackages.crytic-compile: init at 0.1.6
```

asymmetric commented on pkgs/development/python-modules/crytic-compile/default.nix:
```diff
@@ -0,0 +1,39 @@
+      Crytic-compile is a library to help smart contract compilation.
+      It is used in many Crytic tools, e.g. Slither.
+    '';
+    homepage = https://github.com/crytic/crytic-compile;
```
The URL should be double quoted, the URL type is going away.

jerengie replied:
> @jerengie thanks, this looks good! (modulo the changes outlined by @jonringer)
> Would you consider adding yourself to the maintainers? I'm not using this anymore.

Thank you. Sure, I now added myself to the list of maintainers for `slither-analyzer`.

asymmetric commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1552,6 +1552,8 @@ in
 
   crudini = callPackage ../tools/misc/crudini { };
 
+  crytic-compile = with python3Packages; toPythonApplication crytic-compile;
```
Does it make sense to have this in `all-packages`? Seems like it will mostly be used as a dependency of other packages, so not sure users will want to install it directly very often.

jerengie replied:
Thanks for this suggestion.
To be honest, I have no idea if users would like to install `crytic-compile` in a standalone way.
But if desired, it is nice to have the opportunity to do so, isn't it?

FRidh replied:
No we shouldn't pollute the top-level unnecessarily.

asymmetric replied:
@jerengie it will still be available under `python3Packages`, so it can still be installed, just not from the top-level namespace.

@FRidh the manual does say though:

> A distinction is made between applications and libraries, however, sometimes a package is used as both. In this case the package is added as a library to `python-packages.nix` and as an application to `all-packages.nix`.

FRidh replied:
@asymmetric yes but if the author of the PR isn't going to use it like that to begin with, should we then add it just for potential users?

asymmetric replied:
I guess not. Maybe the manual should mention that the package *may* be added to `all-packages.nix`.

asymmetric reviewed (commented):
@jerengie thanks, this looks good! (modulo the changes outlined by @jonringer)

Would you consider adding yourself to the maintainers? I'm not using this anymore.

jerengie committed:
maintainers: add jerengie

jerengie committed:
pythonPackages.crytic-compile: init at 0.1.6

jerengie committed:
pythonPackages.slither-analyzer: fix dependencies

This package can not be build in the current state.
Adding pysha3 and the new crytic-compile package as a dependency fixes this.

FRidh reviewed (changes requested):
- patch code to executable
- remove top-level attribute

jonringer commented:
stale, and conflicts

I'm closing this PR due to lack of activity, if the PR author would like to pursue this matter again, please re-open the PR and address the concerns.
